Best-effort support of Qualified Imports in GHC 9.4

konn opened the pull request:
As discussed in #3473, we cannot provide `import qualified` feature in GHC 9.4 due to the [regression in GHC](https://gitlab.haskell.org/ghc/ghc/-/issues/22130), at least based on the textual information given by error messages.

On the other hand, we can still determine qualified module name from the information of error-range and textual source code.
As I am heavily using `Import Qualified` feature everyday life, so I try to follow this path to tentatively support it in GHC 9.4.
Even though there can be some corner cases that the method implemented in this PR cannot handle correctly, but even a partial support of this could be still helpful.

# TODOs

- [x] Check if this works with multibyte symbols
- ~~(Optional) revive `adding new import item to existing list` along this line~~ This cannot be handled as GHC 9.4 won't give us suggested imports.

konn commented:
It seems that [regression #22130 in GHC](https://gitlab.haskell.org/ghc/ghc/-/issues/22130) removes import suggestion for qualified symbols. This cannot be fixed even with our approach, so I decided to concentrate on reviving `add qualified import` feature in this PR.

konn commented:
I think this PR is ready for review now.

fendor commented on plugins/hls-refactor-plugin/src/Development/IDE/Plugin/CodeAction.hs:
```diff
@@ -1473,7 +1476,7 @@ suggestNewOrExtendImportForClassMethod packageExportsMap ps fileContents Diagnos
 
 suggestNewImport :: DynFlags -> ExportsMap -> Annotated ParsedSource -> T.Text -> Diagnostic -> [(T.Text, CodeActionKind, TextEdit)]
-suggestNewImport df packageExportsMap ps fileContents Diagnostic{_message}
+suggestNewImport df packageExportsMap ps fileContents Diagnostic{_message, ..}
```
I think it would be better if we decide on one style or the other within the same function. Personally, I don't mind either style.

konn replied:
Good catch! I changed it to use RecordWildCards alone in d0e5765af1803fe8fd4fb3ee687336aac6d83fa0.

fendor commented on plugins/hls-refactor-plugin/src/Development/IDE/Plugin/CodeAction.hs:
```diff
@@ -1482,17 +1485,62 @@ suggestNewImport df packageExportsMap ps fileContents Diagnostic{_message}
     L _ HsModule {..} = astA ps
 suggestNewImport _ _ _ _ _ = []
 
+-- tentative workaround for detecting qualification in GHC 9.4
+-- FIXME: We can delete this after dropping the support for GHC 9.4
+extractQualifiedModuleNameFromMissingName :: T.Text -> Maybe T.Text
```
Please, add a small haddock comment.
I had good experience with code examples like this: https://github.com/haskell/hie-bios/blob/master/src/HIE/Bios/Cradle.hs#L748
Gives you an intuition what the function does without having to read it.

Also, perhaps the link to the issue would be great.

michaelpj replied:
:100: for examples

konn replied:
Added in 02cc20110d91e71e55c4a036202daf4f32f8aa69.
Eval plugin didn't work for some reasons, so I evaluated example lines with GHCi manually.

fendor commented on plugins/hls-refactor-plugin/test/Main.hs:
```diff
@@ -1670,10 +1670,11 @@ suggestImportTests = testGroup "suggest import actions"
-      $ test True []          "f = NE.nonEmpty"             []                "import qualified Data.List.NonEmpty as NE"
-    , ignoreForGHC94 "On GHC 9.4 the error message doesn't contain the qualified module name: https://gitlab.haskell.org/ghc/ghc/-/issues/20472"
-      $ test True []          "f = Data.List.NonEmpty.nonEmpty" []            "import qualified Data.List.NonEmpty"
+    , test True []          "f = NE.nonEmpty"             []                "import qualified Data.List.NonEmpty as NE"
+    , test True []          "f = (NE.:|)"             []                "import qualified Data.List.NonEmpty as NE"
+    , test True []          "f = (Data.List.NonEmpty.:|)" []            "import qualified Data.List.NonEmpty"
+    , test True []          "f = (B..|.)"             []                "import qualified Data.Bits as B"
+    , test True []          "f = (Data.Bits..|.)" []            "import qualified Data.Bits"
```
While we are here, can you make sure the indentation looks correct again?

konn replied:
Addressed in 9e1a23097581e05afff2daed660f9f8ccbcb1b5b.

fendor commented on hls-plugin-api/src/Ide/PluginUtils.hs:
```diff
@@ -223,11 +224,27 @@ usePropertyLsp kn pId p = do
+-- | Extracts lines in the text overlapping the given range.
+--
+-- See also: 'extractTextInRange'
+extractOverlappingLinesWithRange :: Range -> T.Text -> T.Text
+extractOverlappingLinesWithRange (Range (Position sl _) (Position el _)) s = newS
```
I am not sure if I like the name, but I can't think of anything else. @michaelpj any ideas?

michaelpj replied:
If it were me, I would be tempted to instead provide `expandToLines :: Range -> Range` and make `extractTextInRange` exact.

fendor replied:
Perhaps `extractLinesInRange`?

fendor replied:
Migrating to Rope would be nice, since we would presumably speed up the formatting code actions for selections :)

michaelpj replied:
It's not necessarily just fine. This is in a code action, and I don't think we have resolve support here. So this may make the code action request grindingly slow on large files on 9.4 if there is an out-of-scope identifier. Maybe we're okay with that but it's potentially a real regression. I'll make a ticket anyway.

michaelpj replied:
https://github.com/haskell/haskell-language-server/issues/3720

fendor replied:
Fair enough.

konn replied:
I followed the suggestion from @michaelpj in b2c5f78513b65b1f0efe2936f1a950b993870450.
Migration to `Rope` (#3720) sounds promising. It seems like a change affecting wider range than the initial purpose of this PR, so I would rather leave it to the future refactoring (by someone else, or myself if the time allows :grin:).

michaelpj replied:
Yes, definitely out of scope for this PR!

fendor reviewed (approved):
LGTM! Thank you very much, this is a great addition!

Small documentation things, otherwise, gtg.

michaelpj commented on plugins/hls-refactor-plugin/hls-refactor-plugin.cabal:
```diff
@@ -86,6 +86,9 @@ library
     , lens
     , data-default
     , time
+    -- FIXME: Only needed to workaround for qualified imports in GHC 9.4
```
Do you want to guard the inclusion of these with a cabal conditional to make it explicit?

fendor replied:
I think that's not worth the trouble. What are a couple of not 100% necessary dependencies that we likely depend on already, in comparison to a cabal flag?

konn replied:
Given that `regex-applicative` is stable and somewhat lightweight dependency, I would leave the dependency as-is.

michaelpj commented on plugins/hls-refactor-plugin/src/Development/IDE/Plugin/CodeAction.hs:
```diff
@@ -1482,17 +1485,62 @@ suggestNewImport df packageExportsMap ps fileContents Diagnostic{_message}
+  , -- tentative workaround for detecting qualification in GHC 9.4
+    -- FIXME: We can delete this after dropping the support for GHC 9.4
+    qualGHC94 <-
+        guard (ghcVersion == GHC94)
```
alternatively we could use some CPP. that would let us only have the dependencies conditionally

fendor replied:
Why CPP when this works just as well? CPP is always akward, to format, to read... I think this is great the way it is.

michaelpj replied:
The only reason would be to make the extra dependencies conditional, but I agree it's probably not worth it.

konn replied:
As I decided to keep dependency for `regex-applicative` as-is, I also left the import unchanged.
I don't like conditional import so much as it also makes `add import` code action behaving awkwardly; e.g. adding import inside CPP conditional block in some cases even when meant as an unconditional import!

konn replied:
Oops, this thread was meant for logic not import - sorry for my misreading. But again, I don't like CPP so much when we can use source-level conditional instead with a little overhead. In this case I expect (but not verified) compilers to optimise away this conditional.

konn commented:
@fendor @michaelpj Thank you for your reviews and sorry for my late response! I had been a bit busy this week. I'm starting responding to the reviews from now.

konn commented on plugins/hls-refactor-plugin/src/Development/IDE/Plugin/CodeAction.hs:
```diff
@@ -1482,17 +1485,62 @@ suggestNewImport df packageExportsMap ps fileContents Diagnostic{_message}
+            | GHC94 <- ghcVersion
+            , isNothing (qual <|> qual')
+            , Just q <- qualGHC94 =
+                mapNotInScope ((q <> ".") <>) thingMissing
```
We need it because the remaining logic assumes `missing` (formerly `thingMissing`) to be qualified name.
I added a note in a8ecc929e39d4a5b66161ce50e340759e8ef39fa.

konn commented:
I think I addressed all the reviews so far. Thank you for your feedbacks! I mark some threads that as resolved with some changes, but please feel free to unresolve threads if you are unsatisfied with changes.

fendor commented on plugins/hls-refactor-plugin/src/Development/IDE/Plugin/CodeAction.hs:
```diff
@@ -1740,6 +1815,11 @@ data NotInScope
     | NotInScopeThing T.Text
     deriving Show
 
+mapNotInScope :: (T.Text -> T.Text) -> NotInScope -> NotInScope
+mapNotInScope f (NotInScopeDataConstructor d) = NotInScopeDataConstructor (f d)
+mapNotInScope f (NotInScopeTypeConstructorOrClass d) = NotInScopeTypeConstructorOrClass (f d)
+mapNotInScope f (NotInScopeThing d) = NotInScopeThing (f d)
```
Is this function used for anything else? If not, we can likely turn it into a local function (e.g. in the where block) and remove the parameter f with the thing it is supposed to do (adding a ".")

konn replied:
Good point! I made it local, changed to more speic name `qualified`, and take qualifier `q` instead of general mapping function `f`.
Addressed in e0eb90a159d0ede5c8b6aab106a79d2d994f1b7f.

fendor reviewed (approved):
LGTM, thank you very much!

Two small comments

konn commented:
Test failure with GHC _9.2_ on Win seems like a genuine regression:

~~~
Test suite func-test: RUNNING...
haskell-language-server
  code actions
    redundant import code actions
      remove solitary redundant imports: FAIL (6.48s)
        test\functional\FunctionalCodeAction.hs:259:
        expected: Nothing
         but got: Just (Command {_title = "Execute Code Action", _command = "4884:hlint:codeActionResolve", _arguments = Just [Object (fromList [("data",Object (fromList [("_uri",String "file:///D:/a/haskell-language-server/haskell-language-server/test/testdata/redundantImportTest/src/CodeActionRedundant.hs"),("_value",Object (fromList [("ignoreHintTitle",String "Use module export list"),("tag",String "IgnoreHint"),("verTxtDocId",Object (fromList [("uri",String "file:///D:/a/haskell-language-server/haskell-language-server/test/testdata/redundantImportTest/src/CodeActionRedundant.hs"),("version",Number 0.0)]))]))])),("diagnostics",Array [Object (fromList [("code",String "Use module export list"),("message",String "Use module export list\nFound:\n  module CodeActionRedundant where\nWhy not:\n  module CodeActionRedundant (\n        module CodeActionRedundant\n    ) where\nan explicit list is usually better\n"),("range",Object (fromList [("end",Object (fromList [("character",Number 0.0),("line",Number 0.0)])),("start",Object (fromList [("character",Number 0.0),("line",Number 0.0)]))])),("severity",Number 3.0),("source",String "hlint")])]),("isPreferred",Bool False),("kind",String "quickfix"),("title",String "Ignore hint \"Use module export list\" in this module")])]})
~~~

https://github.com/haskell/haskell-language-server/blob/8c22b84f897ef6d0c35bc6b53eb7911c478b5473/test/functional/FunctionalCodeAction.hs#L259

https://github.com/haskell/haskell-language-server/actions/runs/5629658777/job/15256519280?pr=3712#step:10:211

It seems rather strange as all other versions of GHC on Win passes tests. 
Does anybody guess the root cause, or is it just a flaky test?

fendor commented:
I presume flaky error, I rerun the failing CI run.

konn commented:
@fendor Now it seems that all `Testing`s are green even with GHC 9.4 on Win. Thanks!

It smees Nix build on ubuntu also failed, but it seems due to the storage shortage. So I suppose everything is OK, right?

fendor commented:
@konn CI and I are happy, but you'll need to rebase the PR 😅
